### DISCUSSION TITLE
Reinitialization on component level

### DISCUSSION THREAD

**Discussion by mysuf** [react-hook-form/react-hook-form#7694] [Q&A] [answered]
**Usecase:**
We have form with fields initialized on component level (`useForm()` without defaultValues) and multiple submit buttons, e.g.:

1. "Submit and navigate to list"
2. "Submit and stay"
3. "Submit and create new one".

Form context and `useForm()` is persistent(mounted) unless user leaves the form route(first submit option from above example). We won't  unnecessary refeshing/rendering => flickering/jumping. 

On POST/PUT backend returns updated form data so if user stays in current context we need to reinitialize whole form with new values (including default values as these are basically backend values).

Calling `useForm()` multiple times has no effect as it probably uses memoization. Currently, we force to update data via `reset(backendData)` inside onSubmit handler which isn't recommended for several reasons according to docs. We do it anyway, because we don't know other way except unmounting whole form context. Two more cons of `reset()` is that:
1. it doesn't reflect rendered components (POST and PUT fields can differ) - so we have to manage shouldUnregister on component level
2. it triggers `reset` event (which can interfere with other plugins listening for this event, f.e. tinymce) even it's not actually meant as true form reset (like through reset button).

**The question:**
 Is there programatic way how to force form hook to reilinitalize whole form on component level (like if you would be calling useForm() for first time) and could be called inside onSubmit handler?

**Comment by Moshyfawn** [comment, marked as answer]
It's a very broad and interesting topic you've brought up in terms of both UX & DX. I'll try to provide some insight on the matter, but I have a couple of questions as well.

### How do you handle multiple different actions for submitting your form?
- Do you access the `event.nativeEvent.submitter.name` in the submit handler and have conditions for the button submitter name to know which action you need to execute?
```js
const onSubmit = async () => {
  await fetch("https://example.come/profile", {
    method: 'POST'
  })
  if (event.nativeEvent.submitter.name === "submit-close") {
    closeForm()
  }
}

<button name='submit-close'>Submit and navigate to list</button>
```
- Do you call your submit handler function `onClick` of every different action button?
```js
<button name='submit-close' onClick={handleSubmit(onSubmitClose)}>Submit and navigate to list</button>
<button name='submit-close' onClick={handleSubmit(onSubmitStay)}>Submit and stay</button>
``` 
### How to reinitialize whole form with new values _(including default values)_
The [`reset` form method](https://react-hook-form.com/api/useform/reset/#main) is the method to reset your form defaults and field values. You can look into the `reset`'s API options like `keepSubmitCount` or `keepValues` to determine whether or not the form has already been submitted or keep your form's field values but reset their defaults, if the BE data hasn't been modified and correlates with the already filled out values.
### Calling `useForm()` multiple times has no effect
I don't completely understand what do you mean by "calling the form hook multiple times". In general, React doesn't allow for conditionally calling hooks.
Also, calling `useForm` multiple times would result in multiple different independent form instances.
### Updating data via `reset` form method
It's the only correct form method to use in your use-case. Since [the 7.22.0 release](https://github.com/react-hook-form/react-hook-form/releases/tag/v7.22.0), you don't need to create a reset side-effect inside the `useEffect` hook like
```js
useEffect(() => {
  if (formState.isSubmitSuccessful) {
    reset(newDefaultValues)
  }
}, [reset, formState.isSubmitSuccessful, newDefaultValues])
```
It's allowed to call reset inside the `onSubmit` like
```js
const onSubmit = async (data) => {
   const newDefaultValues = await submitData(data)
   reset(newDefaultValues)
}
```
P.S. I'll see if we need to update the docs accordingly
### `reset` doesn't reflect rendered components
Can you elaborate more on what do you mean here? There's a `shouldUnregister` prop on `useForm` hook, that will unregister the hidden inputs without you calling [the `unregister` form method](https://react-hook-form.com/api/useform/unregister) on the component level.
### `reset` triggers the reset form event
Can you provide an example _(a codesandbox would be awesome)_ if the undesired side-effect of RHF calling emitting the `reset` event on the `reset` form method called?
### Is there programmatic way how to force form hook to reinitialize whole form on component level _(like if you would be calling `useForm` for first time)_ and could be called inside `onSubmit` handler?
The [`reset` form method](https://react-hook-form.com/api/useform/reset/#main) is that programmatic method that can help you reset your form defaults and values. As mentioned earlier, you can call it inside the `onSubmit` since [the 7.22.0 release](https://github.com/react-hook-form/react-hook-form/releases/tag/v7.22.0).

**Reply by Moshyfawn** [reply]
There's, also, [the resetField form method](https://react-hook-form.com/api/useform/resetfield/#main) that allows for resetting a single form field, as well. But I'm not sure that's the right fit for reseting multiple form fields along with the form's state

**Comment by mysuf** [comment]
Thank you for very descriptive answer.

> How do you handle multiple different actions for submitting your form? Do you access the event.nativeEvent.submitter.name in the submit handler?

Yes. We obtain the name of action from nativeEvent attached to react-hook-form's `handleSubmit`. So we use only one method and post-submit action is handled inside.

> Calling useForm() multiple times has no effect

I mean re-rendering component where `const formMethods = useForm();` is placed. Like you said, it's hook so it probably keeps its context like other hooks, e.g. `useEffect`, but since `useForm` doesn't accept any dependencies like other hooks do, it's cached unless it's unmounted.

> reset triggers the reset form event

 I haven't been able to replicate this behaviour in sandbox from home so if I'll face this issue again and find it out somehow related to react-form-hook, I'll provide more information.

> How to reinitialize whole form with new values (including default values)

If calling `reset` inside submit handler is fine and `shouldUnregister` is the way how to reflect changes on component level (seems working so far), feel free to close this as resolved.

Thanks again

**Reply by Moshyfawn** [reply]
Thank you for your clarifications!
> I haven't been able to replicate this behaviour in sandbox _(about reset event)_

RHF tries to obtain a `ref` of some element registered form element to get it's closest form element parent and call it's reset method.
```js
isHTMLElement(fieldReference) &&
  fieldReference.closest('form')!.reset();
```
If you don't provide a `ref` callback form `register` or `field.ref` from `<Controller />` / `useController`, RHF won't be able to call the native HTML reset on the form element. See [this CSB](https://codesandbox.io/s/eloquent-https-29gud?file=/src/App.tsx): notice `ref={null}` to override the `register` form method spreading the `ref` callback into the input. If you remove `ref={null}`, you'll see an alert `onReset` callback
> If calling reset inside submit handler is fine and shouldUnregister is the way how to reflect changes on component level

Yep, it's the way to handle form's reset.
Feel free to ask any further questions, if you get any ;)

**Reply by Moshyfawn** [reply]
Oh, and apparently, if you provide defaultValues to `reset` as an argument, the form reset event won't be fired, huh..
@bluebill1049 can you help me understand what's the logic behind this condition in `reset`? `isWeb` is self-explanatory, but why the `undefined` check and if no new default values, try calling the form's reset?
```js
if (isWeb && isUndefined(formValues))
...
// Try calling .reset() on the closest form element 
...
```

**Reply by bluebill1049** [reply]
if you don't supply a `defaultValues`, then react hook form will invoke the native reset API, because we don't know what's the `defaultValues` for the reset. It's already recommended to provide values for the reset API.


```tsx
const {reset, register} = useForm()

<input {...register('data')} />

reset() // can't predict defaultValues

// each input has a defaultValue, this is especially important for controlled component
reset({
  data: ''
})
```

**Reply by Moshyfawn** [reply]
I think I understand now, thx, @bluebill1049!
So, is it not recommended to set `type="reset"` on a button and call `reset` form method `onClick={() => reset(defaultValues)}?

**Reply by bluebill1049** [reply]
It depends:

- **If you are using a full uncontrolled form without any defaultValue**

You just want to reset your form to an "empty" state without any value. `type="reset"` is the most ideal usage, in fact, you probably don't need to use the `reset` API at all.

- **If you have `defaultValues`, controlled inputs**

Then you should always supply `defaultValues` for the reset API, in fact, I am thinking of deprecating the native reset API in the next major version.  

include this information in the doc

**Reply by Moshyfawn** [reply]
I'm just trying to think: if I click the `type="reset"` button and put the `reset` form method inside the `onClick` callback, the form's reset event will fire anyway. Can it create an unwanted behavior?
[An example CSB](https://codesandbox.io/s/keen-banzai-vdw98?file=/src/App.tsx)

**Reply by bluebill1049** [reply]
I don't think you need to combine `reset` with `type="reset"`, probably better to use either.  What sort of usage would require both?

> Can it create an unwanted behavior?

It would not unmount inputs.

**Reply by Moshyfawn** [reply]
> What sort of usage would require both?

I have a generic form component that has both reset and submit action buttons with their types reset and submit types accordingly. There's no check whether or not the form contains controlled inputs, so all forms have those action buttons with types. `onSubmit` does it's async `POST` / `PUT` / whtvr and there's a condition, if there're some defaults returned from an API and the form isn't supposed to be closed `onSubmit`, then `reset` with those API values, as they can be formatted on BE or have some additional fields prepended.

I'm not saying it's necessarily my use-case, meaning it's **not** a requirement to always have the `type="reset"` on the button. It might be the case of me going and fixing the form according to the RHF expected usage. Just an edge-case I wanna understand how to go about

**Reply by bluebill1049** [reply]
I see. when you reset the form, do you require any `formState` or values update? If not, then you can simply use `type="reset"`, else i think it's safer to use `reset` API from hook form.
